### DISCUSSION TITLE
Gradual color shifting

### DISCUSSION THREAD

**Discussion by guillaumeboehm** [loqusion/hyprshade#28] [Ideas]
Hey there,

Nice project! I'm definitely going to follow the progress.

I've seen on the readme that gradual color shifting is not yet supported, do you have an idea of about when you intend to tackle this?

I've looked around the code to see if I couldn't do it somehow and I got somewhat close with the new mustache configs. I thought that a potential short term solution might be to generate separate shaders when a shader is redefined in the `config.toml` with different variable values.
For example having the `temperatureStrength` as a mustache variable and setting X shaders in the `config.toml` with increasing `temperatureStrength`. It's a bit tedious but better than creating X shaders by hand and setting them in the config xD

What do you all think?

**Comment by loqusion** [comment]
Currently, the only code that is time-aware is [`Schedule`](https://github.com/loqusion/hyprshade/blob/542d03578616f3db0467ce3a944566677e780add/src/hyprshade/config/schedule.py#L19), meaning the rest of the code either make no decisions based on time or they defer to `Schedule`. In its current implementation, `Schedule` will construct a [`Shader`](https://github.com/loqusion/hyprshade/blob/542d03578616f3db0467ce3a944566677e780add/src/hyprshade/shader/core.py#L88) that is [passed](https://github.com/loqusion/hyprshade/blob/542d03578616f3db0467ce3a944566677e780add/src/hyprshade/config/schedule.py#L28) the entire configuration object derived from `hyprshade.toml`, and that `Shader` naively [plucks](https://github.com/loqusion/hyprshade/blob/542d03578616f3db0467ce3a944566677e780add/src/hyprshade/config/core.py#L27) the first item in `shades` that matches its `name`.

TL;DR multiple items in `shades` with the same name are not currently supported.

Rather than placing the burden on the user to "implement" color shifting by manually specifying each time the `blue-light-filter` will be reactivated, the ideal scenario would be a special handling of `blue-light-filter` that essentially does that work for the user. I'm not sure when I'd get around to implementing it however, since right now my priority is stabilizing the features that will be released in v4.0.0.

**Reply by guillaumeboehm** [reply]
I see, maybe I'll try to code something while you get 4.0.0 ready. Python is not my goto language so we'll see lol.

Would you feel better with a blocking execution of hyprshade when a gradual shift is available or reusing the current schedule way and set a bunch of timers to act as a gradual shift?

I was thinking of adding a `gradual_shift_duration` hyprshade.toml variable that would automatically set a `gradualPercentage` glsl variable if the shader is a template, and just ignore the variable if it is not. So anyone could implement their shader with the `gradualPercentage` variable used properly to shift gradually.

**Reply by guillaumeboehm** [reply]
I did a POC PR here https://github.com/loqusion/hyprshade/pull/29 reusing the systemd timers. It's working fine but it's maybe putting some unwanted load on systemd tasks. And it is not possible(I think or at least probably not smart) to use shift steps smaller than 1 second so small shift durations are quite visible.

I feel like having a blocking execution running would allow for smoother shifts when the shift duration is small, and less work for systemd. Although I didn't look at the hyprctl code at all so I don't even know if that's possible.

I think I'll make a blocking solution POC in an other branch to see.

**Reply by guillaumeboehm** [reply]
Just finished the blocking POC here #30, I think it's less glued together than the systemd solution. If one of the implementations interests you for after the next major, you can tell me and I'll clean up the PR.
